### DISCUSSION TITLE
ScreenViewport query?

### DISCUSSION THREAD

**Discussion by JohnMajerle** [iTwin/itwinjs-core#5265] [Q&A] [answered]
Given a ScreenViewport, how would I construct a query for the visible elements?

**Comment by pmconne** [comment]
That depends upon how you define "visible".
Try starting with [Viewport.queryVisibleFeatures](https://www.itwinjs.org/reference/core-frontend/views/viewport/queryvisiblefeatures/).

**Comment by JohnMajerle** [comment]
Hey Paul,

Hmm. I guess I define visible as something that I can see in the ViewPort. :-)

My attempt below always returns true even for areas where I have intentionally zoomed into empty space and see only the black background.  If I instead modify it to get a count of features I always get a large positive number.  I get the same result if I don't pass a rect value to QueryVisibleFeatureOptions.

Are there additional options I need to set so that I detect zero elements for a blank region?

Thanks,

John 


  private static containsFeatures(vp: ScreenViewport): boolean {

    let result : boolean = false;

    let options: QueryVisibleFeaturesOptions = { 
      source: "screen",
      rect: vp.viewRect
    };

    vp.queryVisibleFeatures(options, (features) => {
      for (const feature of features) {
        if (feature.iModel === vp.iModel && Id64.isValid(feature.elementId)) {
          result = true;
          break;
        }
      }
    });    

    return result;
  }

**Reply by pmconne** [reply]
Try excluding any elementIds for which `Id64.isTransient` returns `true` or `Id64.isValidId64` returns `false`.
If problems persist, please change your callback to populate a `Set` of element Ids instead of breaking on the first feature. Then after the loop exits, output `JSON.stringify(Array.from(yourSet))` and paste the result here.

**Comment by JohnMajerle** [comment]
Hey Paul,

Filtering out transients did the trick.  See modified function below.

What exactly is a transient element?

--John


  private static containsFeatures(vp: ScreenViewport): boolean {

    let result : boolean = false;

    let options: QueryVisibleFeaturesOptions = { 
      source: "screen",
      rect: vp.viewRect
    };

    vp.queryVisibleFeatures(options, (features) => {
      for (const feature of features) {
        if (Id64.isTransient(feature.elementId))
          continue;

        if (feature.iModel === vp.iModel && Id64.isValid(feature.elementId)) {
          result = true;
          break;
        }
      }
    });    

    return result;
  }

**Reply by pmconne** [reply]
Anything inside a viewport that can be interacted with needs a unique Id. For elements, that Id is the persistent element Id. For other, non-persistent things - like pickable decorations and reality models - we assign transient Ids, which begin with the prefix `0xffffff`. That prefix is reserved for transient (non-persistent) things so transient Ids never conflict with persistent Ids.

**Comment by JohnMajerle** [comment, marked as answer]
Cool.  Thanks for the explanation.
